h/api/db/types.py

```diff
@@ -65,6 +66,79 @@ def process_result_value(self, value, dialect):
         return _get_urlsafe_from_hex(hexstring)
 
 
+class MutableList(Mutable, list):
+
+    """
+    A list subclass that persists in-place changes.
+
+    If you use a normal list as an attribute of an ORM class then in-place
+    changes to the list won't be persisted to the session, so use this class
+    instead.
+
+    To use this class wrap another list-like type in this class's as_mutable()
+    class method when declaring a column:
+
+        my_array_column = sa.Column(
+            MutableList.as_mutable(pg.Array(sa.UnicodeText)))
+
+    See: http://docs.sqlalchemy.org/en/rel_1_0/orm/extensions/mutable.html
+    """
+
+    @classmethod
+    def coerce(cls, key, value):
+        if not isinstance(value, MutableList):
+            if isinstance(value, list):
+                return MutableList(value)
+            return Mutable.coerce(key, value)
+        else:
+            return value
+
+    def __setitem__(self, idx, value):
```

Comment:
It wouldn't be too hard, I don't think, to write a fairly concise set of tests for this class. Something like the following, perhaps:

```python
@mock.patch.object(MutableList, 'changed')
@pytest.mark.parametrize('start,end,retval,op,args', (
    (['foo'], ['bar'], None, '__setitem__', (0, 'bar')),
    ([1,2,3], [1,4,5], None, '__setslice__', (1, 3, [4, 5])),
    ...
))
def test_mutable_list_operation(start, end, retval, op, args, changed):
    val = MutableList(start)
    result = getattr(val, op)(*args)
    
    assert result == retval
    assert val == end
    changed.assert_called_once_with()
```